### PR TITLE
feat(AWS HTTP API): Add support for custom Lambda authorizers

### PR DESCRIPTION
**Pull request by pgrzesik**
Adds support for custom Lambda authorizers for HTTP API. There is a slight change to original proposal, instead of `name` and `arn` I used `functionName` and `functionArn` because `name` was already reserved for overriding authorizer name. 

Additionally it includes a small fix to `generatePayload` to a bug that was exposed after adding authorizer functions to `httpApi` fixture. 

Addresses: #8210

### CODE REVIEW

File: lib/plugins/aws/package/compile/events/httpApi.js [hunk 1]
```diff
@@ -88,6 +89,7 @@ class HttpApiEvents {
                     },
                     name: { type: 'string' },
                     scopes: { type: 'array', items: { type: 'string' } },
+                    type: { type: 'string', regexp: new RegExp('^(REQUEST|JWT)$', 'i').toString() },
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/httpApi.js:92]: I would stick to single case support, and define it here with _enum_.

It think there's no clear benefit in supporting case insensitivity here, while it's  definitely confusing and implementation wise error-prone.

**Review by medikoo — COMMENTED** [review]
@pgrzesik looks great. Still I would simplify input forms we support for authorizer type

File: lib/plugins/aws/package/compile/events/httpApi.js (hunk 1 above)
Comment (pgrzesik) [reply lib/plugins/aws/package/compile/events/httpApi.js:92]: I was thinking about it for a moment and I've went with this approach to be consistent with authorizers for `http` event, where we support case insensitvity and even examples in documentation use both notations. Do you think it would be better to introduce that change which will make `httpApi` implementation a bit more consistent while sacrificing consistency between `http` and `httpApi` ?

Comment (medikoo) [reply lib/plugins/aws/package/compile/events/httpApi.js:92]: I think that case insensitivity in case of `http` was not great decision. It also predated config based validation.

I would not look back here, same as when deciding on other parts of the config we didn't look back (e.g. in HTTP API it's `provider.httpApi.name` while with Rest API it's `provider.apiName`)

Also if user will migrate from `http` and start to use upper case with `httpApi` it'll get very meaningful error message, so I don't think it should be a problem.

Comment (pgrzesik) [reply lib/plugins/aws/package/compile/events/httpApi.js]: :+1: changed

File: lib/utils/analytics/generatePayload.js [hunk 2]
```diff
@@ -98,12 +98,15 @@ module.exports = async (serverless) => {
...
       },
       plugins: serviceConfig.plugins ? serviceConfig.plugins.modules || serviceConfig.plugins : [],
-      functions: Object.values(serviceConfig.functions).map((functionConfig) => ({
-        runtime: functionConfig.runtime || defaultRuntime,
-        events: functionConfig.events.map((eventConfig) => ({
-          type: Object.keys(eventConfig)[0] || null,
-        })),
-      })),
+      functions: Object.values(serviceConfig.functions).map((functionConfig) => {
+        const functionEvents = functionConfig.events || [];
```
Comment (medikoo) [inline lib/utils/analytics/generatePayload.js:102]: Why this change is necessary? Normally `functionConfig.events` is guaranteed internally

**Review by medikoo — COMMENTED** [review]
Thanks @pgrzesik looks very good. It's just not clear to me why change in `generatePayload` was introduced.

File: lib/utils/analytics/generatePayload.js (hunk 2 above)
Comment (pgrzesik) [reply lib/utils/analytics/generatePayload.js:102]: It was due to a quirky test setup of https://github.com/serverless/serverless/blob/bf83f3c406b576599c7f37a37e583ac6a530b1d3/test/unit/lib/utils/analytics/generatePayload.test.js#L18
In this case, the existence of `events` as an array is not ensured on `functions`. I've decided to patch it in a way that will ensure that `generatePayload` will default to an empty list in such edge cases.

Comment (medikoo) [reply lib/utils/analytics/generatePayload.js:102]: hmm.. but it was more an issue in _test_ than in tested module. 

Technically `serverless` instance guarantees that that every function has `events` array, and in such case we should not introduce a noise with such checks (same as we assume that all values match configuration schema)

Comment (medikoo) [reply lib/utils/analytics/generatePayload.js:102]: However, as I think of it. In future we may loose that guarantee. Especially when we will pass  a _configuration input_ directly (by schema, `events` is not a required property).

So after all I think it's fine to add such check here

**Review by medikoo — DISMISSED** [review]
Great work @pgrzesik  👍
